src/filesystem.cpp

```diff
@@ -856,4 +853,207 @@ bool FileInputStream::read(byte *buffer, unsigned size)
+        {
+            localpath.insert(0, localseparator);
+        }
+    }
+    localpath.insert(0, additionalPath.localpath);
+}
+
+void LocalPath::trimNonDriveTrailingSeparator(const FileSystemAccess& fsaccess)
+{
+    if (localpath.size() >= fsaccess.localseparator.size()
+        && !memcmp(localpath.data() + (int(localpath.size()) & -int(fsaccess.localseparator.size())) - fsaccess.localseparator.size(),
+            fsaccess.localseparator.data(),
+            fsaccess.localseparator.size()))
+    {
+        // ok so the last character is a directory separator.  But don't remove it for eg. F:\ on windows
+        #ifdef WIN32
+        if (localpath.size() > 2 * fsaccess.localseparator.size() && !memcmp(localpath.data() + localpath.size() - 2 * fsaccess.localseparator.size(), L":", fsaccess.localseparator.size()))
+        {
+            return;
+        }
+        #endif
+
+        localpath.resize((int(localpath.size()) & -int(fsaccess.localseparator.size())) - fsaccess.localseparator.size());
+    }
+}
+
+bool LocalPath::findNextSeparator(size_t& separatorBytePos, const FileSystemAccess& fsaccess) const
+{
+    for (;;)
+    {
+        separatorBytePos = localpath.find(fsaccess.localseparator, separatorBytePos);
+        if (separatorBytePos == string::npos) return false;
+        if (separatorBytePos % fsaccess.localseparator.size() == 0) return true;
+        separatorBytePos++;
+    }
+}
+
+bool LocalPath::findPrevSeparator(size_t& separatorBytePos, const FileSystemAccess& fsaccess) const
+{
+    for (;;)
+    {
+        separatorBytePos = localpath.rfind(fsaccess.localseparator, separatorBytePos);
+        if (separatorBytePos == string::npos) return false;
+        if (separatorBytePos % fsaccess.localseparator.size() == 0) return true;
+        separatorBytePos--;
+    } 
+}
+
+size_t LocalPath::getLeafnameByteIndex(const FileSystemAccess& fsaccess) const
+{
+    // todo: take utf8 two byte characters into account
+    size_t p = localpath.size();
+    p -= fsaccess.localseparator.size() % 2; // just in case on windows
+    while (p && (p -= fsaccess.localseparator.size()))
+    {
+        if (!memcmp(localpath.data() + p, fsaccess.localseparator.data(), fsaccess.localseparator.size()))
+        {
+            p += fsaccess.localseparator.size();
+            break;
+        }
+    }
+    return p;
+}
+
+bool LocalPath::backEqual(size_t bytePos, const string& compareTo) const
+{
+    auto n = compareTo.size();
+    return bytePos + n == localpath.size() && memcmp(compareTo.data(), localpath.data() + bytePos, n);
+}
+
+bool LocalPath::backEqual(size_t bytePos, const LocalPath& compareTo) const
+{
+    auto n = compareTo.localpath.size();
+    return bytePos + n == localpath.size() && memcmp(compareTo.localpath.data(), localpath.data() + bytePos, n);
+}
+
+LocalPath LocalPath::subpathFrom(size_t bytePos) const
+{
+    return LocalPath::fromLocalname(localpath.substr(bytePos));
```

Comment:
Should we check here that `bytePos <= localpath.size()`? It depends on the usage of this function, but if the condition fails, `string::substr()` will throw out-of-range exception.

Reply:
I was thinking that once it was clear where all the usages of functions like this that need to know indexes into our wrapped string are, we would figure out a better public interface for LocalPath that would mean indexes are not needed, so the interface is independent of the internal representation.   In the meantime, this function should only be called in ways that the code was called previously - which is to say, with a bytePos from getLeafnameByteIndex() or lastpartlocal().    Also those two should be consolidated, but looking at them now they are slightly different...  and without wanting to complicate the PR or risk introducing subtle issues, it might be better to keep this as is in this (already very large) PR, which is performing the same string calculations as before, and make that adjustment in at future PR dedicated to it?  I'll make a note in that new PR for adjusting LocalPath.  thanks!